### ISSUE TITLE
Regex Find can cause VS Code to Freeze

### ISSUE THREAD

**Issue by Jason3S** [Microsoft/vscode#30874]
- VSCode Version: 1.14.1
- OS Version: Mac OSX Sierra 10.12.5 (16F73)

Steps to Reproduce:

1. `code --disable-extensions`
1. *File->New File*
1. Paste file from here: [snippets.py](https://raw.githubusercontent.com/GoogleCloudPlatform/python-docs-samples/master/auth/http-client/snippets.py) (it happens with other files, but this is a simple example)
1. Edit->Find  `"""(.*?\n?)+?"""` regex
1. Try to remove the last `?` from the regex.
1. The editor freezes at this point
1. Code Helper goes to 100% CPU.

Result:
![image](https://user-images.githubusercontent.com/3740137/28280643-bcdb1b42-6b24-11e7-81af-649bca4f45c9.png)

Reproduces without extensions: Yes

**Comment by Jason3S** [comment]
It also happens with: `1.15.0-insider`

**Comment by Jason3S** [comment]
I close it by accident. This is a real issue.

**Comment by rebornix** [comment]
Really good catch!

**Comment by rebornix** [comment]
This is an interesting JavaScript regex issue. If you run the same regex execution against the same content in browser, you can run into the same 100% CPU problem. Regex101 fails gracefully in a good timeout, we may want to do the same.

**Comment by roblourens** [comment]
@rebornix it probably only impacts the search viewlet when the file is open, right? I think Rust's regex engine is not susceptible to this. So it would be fixed by any fix to the editor.

**Comment by rebornix** [comment]
@roblourens yes exactly as you said. It should just be a fix in the editor side.

**Comment by Spown** [comment]
Another case:

![](https://drive.google.com/uc?id=1g10I3hHpQ-yJcDI2_YDMg-dBEjEInX_-)

Steps:

1. Have a file with content
```
t({phrase: 'mail_text_q', locale: 'ru-RU'}, mailCtx)+'\n'+
t({phrase: 'mail_text_m', locale: 'ru-RU'}, mailCtx)+'\n'+
```
2. call search widget
3. set to RegEx
4. input ``(.*+'\n'+``
5. try to add ``)`` after the third character
```
(.*
   ^
```

**Comment by pascallaprade-beslogic** [comment]
I have another case of a regex search that hangs VS Code: `/\*(.*\n?)*\*/`.

- Pasting the regex immediately hangs VS Code, and a window appears after a few seconds, suggesting to reopen the window.
- Typing the regex manually will hang when typing the fourth `*` character specifically (I reproduced it many times to make sure it was this specific character hanging), so: `/\*(.*\n?)*\` works, then as you type the `*` right after that, it hangs.

This is on VS Code 1.21.1, on Windows 10, both x64.

**Comment by ca1ek** [comment]
I found an another one that hangs.  `(.*)\(\d*\`

- VSCode Version: 1.22.2
- OS Version: Windows 10 1709

I attempted to turn a rather large HTML file (577KB) into a CSV containing only data from that file. I can't share that file due to privacy reasons.

1. Press CTRL+H for search and replace 
2. Enable regex and type in `(.*)\(\d*\` and possibly punch in another space after it

There's about 25% CPU activity after it hangs, but I've been waiting a couple of minutes now. A regex shouldn't ever hang the whole editor.

Does this issue occur when all extensions are disabled?: Yes

**Comment by gtzinos** [comment]
Hello all, 

try to disable the new feature `search.usePCRE2` from settings.

I think is better, search works sometimes.

**Comment by stefanJi** [comment]
![](https://ws1.sinaimg.cn/large/005EDc3Bgy1g0v6h6emqzj30sr04lju2.jpg)

Input order:
1. input `\s19:.*`
2. insert `.` to start
3. vscode freezed

**Comment by stefanJi** [comment]
> ![](https://camo.githubusercontent.com/8da26649a913c73402d62db2287e6aa2219d1d30/68747470733a2f2f7773312e73696e61696d672e636e2f6c617267652f3030354544633342677931673076366836656d717a6a3330737230346c6a75322e6a7067)
> 
> Input order:
> 
> 1. input `\s19:.*`
> 2. insert `.` to start
> 3. vscode freezed

Vscod Version: 1.31.1
Os: mac 10.14.3

**Comment by jtakalai** [comment]
Curious that it's such a problem since other editors like Sublime don't have a problem with highlighting regex matches... Why is newline such an issue? Shouldn't it be just another character?

**Comment by hobpet** [comment]
> Curious that it's such a problem since other editors like Sublime don't have a problem with highlighting regex matches... Why is newline such an issue? Shouldn't it be just another character?

Actually I've tested my search in Sublime and there I have the following messages (but no crash!)

```ERROR: Regex exhausted stack searching file```
```ERROR: Regex complexity too high searching file```

**Comment by emilygdavis** [comment]
I am running into this, too, for Find in a single file (not the global find/replace). I'm encountering the issue simply by using a pipe character (`|`) in my RegEx, when I have not yet populated characters after the pipe character. If I try enter something like
```
element|
```
(eventually intending to type `element|getElement`), VSCode freezes up, before I've even hit "Enter" to do the Find.

This is on VSCode version 1.39.2, on Max OSX Mojave (v10.14.6).

**Comment by Fadavvi** [comment]
Version: 1.41.1 (user setup)
Commit: 26076a4de974ead31f97692a0d32f90d735645c0
Date: 2019-12-18T14:58:56.166Z
Electron: 6.1.5
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Windows_NT x64 10.0.18363

and still 

![VSCodeProblem](https://user-images.githubusercontent.com/21213484/71304249-71a90880-23d9-11ea-98f9-3d8cdb21e102.gif)

**Comment by roblourens** [comment]
@Fadavvi please open a new issue, I can't repro it. I'm curious whether it happens without the "whole word" toggle enabled, or in a shorter file.

**Comment by Fadavvi** [comment]
> 
> 
> @Fadavvi please open a new issue, I can't repro it. I'm curious whether it happens without the "whole word" toggle enabled, or in a shorter file.

OK. I'll open a new issue. It happen only on some special files. I can't find similarity of files!

**Comment by jmlowenthal** [comment]
> I am running into this, too, for Find in a single file (not the global find/replace). I'm encountering the issue simply by using a pipe character (`|`) in my RegEx, when I have not yet populated characters after the pipe character. If I try enter something like
> 
> ```
> element|
> ```
> 
> (eventually intending to type `element|getElement`), VSCode freezes up, before I've even hit "Enter" to do the Find.
> 
> This is on VSCode version 1.39.2, on Max OSX Mojave (v10.14.6).

+1 to this on VSCode 1.45.0 on Linux x64 5.6.0-1-amd64 (Debian)

**Comment by Artemis21** [comment]
Very simple regex to reproduce: `(x|)`

**Comment by jmlowenthal** [comment]
Looks to be fixed in 1.45.1 for me. @Artemis21 can you repro in 1.45.1?

**Comment by ndbroadbent** [comment]
I'm using VS Code `1.45.1`, and I've just found another regex that will cause VS Code to freeze: `"([^]+\n)+"`

I've reproduced this with some minimal examples: https://gist.github.com/ndbroadbent/bb5e246b05488296576ee3dc39d40e38 

(I've tested these with `code --disable-extensions` to disable all extensions.)

One example freezes my computer for about 1-2 seconds. The other example just has a few extra lines, but it looks like it causes a huge exponential increase in complexity and freezes VS Code forever.

It would be a really good idea to add a simple timeout so that we don't accidentally freeze the editor.

These error messages from Sublime would be even better:

```
ERROR: Regex exhausted stack searching file
ERROR: Regex complexity too high searching file
```

Thanks!

**Comment by IllusionMH** [comment]
This https://twitter.com/schuay/status/1325715124930420736 looks promising for current case, but wonder if it will be applicable.

Details in Chromium bug tracker: [Issue 10765: Tracking bug: Non-backtracking regexp execution](https://bugs.chromium.org/p/v8/issues/detail?id=10765)

**Comment by tombogle** [comment]
I wrote a fairly complex regex specific to my data, and also got this hang. Even if the regex is ill-formed or just really nasty to where it executes inefficiently, there needs to be a way to cancel the find (which kicks off automatically in the background as you're entering the search string).

**Comment by gustavoteixeiradev** [comment]
I am having this issue when applying the following regex:
`(added )+(\w+)+( output/)+(\w+)+(.png)`
in a file that contains 2000 lines with thius content:
```
added Qmb46Sdqb8JjMEGMXbQa2QTDukEz8Sa5jTLRFYqqxAxqn4 C:\\repositories\\kryptodevelopers\\output\\chunk1/995.png
added QmVXeSf9Matf4FCNkDcAFpUDQNUC4sDKnoJH3KJWUZTTWM C:\\repositories\\kryptodevelopers\\output\\chunk1/996.png
added QmQznUdyBTwMrn1ZhAGz7nK7mQVEKy7TDFT2aycDKEgLNb C:\\repositories\\kryptodevelopers\\output\\chunk1/997.png
added QmWx7ZtP7Gm4ASe7J4HVNQqmXRCuyvU6qNkBC77x59dVtA C:\\repositories\\kryptodevelopers\\output\\chunk1/998.png
added QmcFWFLo1ysNhKQeGBRUrhkmdHQ1heozGR1G4WSczGGLbV C:\\repositories\\kryptodevelopers\\output\\chunk1/999.png
```


These are the versions:
![image](https://user-images.githubusercontent.com/10344032/141305509-f65e070a-0c97-4e6a-b942-dbc66b5dedc7.png)


Pressing CTRL+F and inputing the regex, freezes the IDE and the only way to get it back to work is restarting it.

**Comment by IllusionMH** [comment]
Because you shouldn't use `(\w+)+` which is explained [above](https://github.com/microsoft/vscode/issues/30874#issuecomment-315861140), leave one `+` (think inside of parentheses).

**Comment by awidjaja** [comment]
> Another case:
> 
> ![](https://camo.githubusercontent.com/1fab818be57927cbac6a0930b5d39b8cf3a1abb077e89b7cd4f8b22fd1683cc9/68747470733a2f2f64726976652e676f6f676c652e636f6d2f75633f69643d316731304933684870512d794a634449325f59444d672d6442456a45496e585f2d)
> 
> Steps:
> 
> 1. Have a file with content
> 
> ```
> t({phrase: 'mail_text_q', locale: 'ru-RU'}, mailCtx)+'\n'+
> t({phrase: 'mail_text_m', locale: 'ru-RU'}, mailCtx)+'\n'+
> ```
> 
> 2. call search widget
> 3. set to RegEx
> 4. input `(.*+'\n'+`
> 5. try to add `)` after the third character
> 
> ```
> (.*
>    ^
> ```

I wonder if after 3 years there is a solution for the possessive operator like .*+ ?

**Comment by Jason3S** [comment]
@awidjaja,
Please try your expression on a site like https://regex101.com/ .
If you want to match against `(` or `+`, they need to be escaped.

Like this:
```
\(.*?\)\+'\\n'\+
```

**Comment by awidjaja** [comment]
Hi @Jason3S , vscode crash each time a [multi line](https://code.visualstudio.com/updates/) search fails to find a match, possibly because of regex backtracks hell.

Following the example from the included link, if there is no matching `<body>` tag, vscode will stop functioning and you have to close it and manually cancel vscode process in task manager, risking your un-saved work to be lost.

> [How Possessive Quantifiers Work](https://www.regular-expressions.info/possessive.html)
Like a greedy quantifier, a possessive quantifier repeats the token as many times as possible. Unlike a greedy quantifier, it does not give up matches as the engine backtracks. With a possessive quantifier, the deal is all or nothing. You can make a quantifier possessive by placing an extra + after it. * is greedy, *? is lazy, and *+ is possessive. ++, ?+ and {n,m}+ are all possessive as well. Possessive quantifiers are supported by JGsoft, Java, and PCRE.
>
> Possessive quantifiers allow your regex to fail faster. In the above example, when the closing quote fails to match.
Technically, possessive quantifiers are a notational convenience to place an [atomic group](https://www.regular-expressions.info/atomic.html) around a single quantifier. All regex flavors that support possessive quantifiers also support atomic grouping. But not all regex flavors that support atomic grouping support possessive quantifiers. With those flavors, you can achieve the exact same results using an atomic group. Atomic grouping is supported by most modern regular expression flavors, including the JGsoft flavor, Java, PCRE, .NET, Perl, Boost, and Ruby. Most of these also support possessive quantifiers, which are essentially a notational convenience for atomic grouping.

Currently, both `atomic group (?>X*)` and `possessive identifier X*+` doesn't work in vscode search.

I was hoping there is already a backlog to make the multi line search really work without causing vscode to crash after 3 years had gone. Hope this clarifies.

**Comment by Jason3S** [comment]
@awidjaja,
That would be a feature request. It might be worth it to open a new feature request and mention that it could be a solution to this issue. 

My guess is that VS Code uses the built in RegExp Engine from JavaScript, it does not support possessive qualifiers.
One suggestion would be to use the `oniguruma` engine that VS Code uses to colorize code.

**Comment by awidjaja** [comment]
Indeed, [javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) has not support both possessive quantifier and atomic grouping. 

I thought vscode regex was based on PCRE or ripgrep engine?

I tested some regex using named group capture that is supported by javascript. vscode allows it but there is no way to reference the group in the replace box, neither ${name} nor \k<name> work. Are you sure the regex was based on javascript?

**Comment by IllusionMH** [comment]
Editor's content stored in JS and therefore JS(V8) regexp engine is used for search in editor.

If you use Find in files - ripgrep (PCRE2 compliant) is used for unopened files (opened files currently always use JS engine).

As for replace - correct syntax would be `$<name>` but it might be just escaped before passing to replace and should be investigated separately.

**Comment by awidjaja** [comment]
> Editor's content stored in JS and therefore JS(V8) regexp engine is used for search in editor.
> 
> If you use Find in files - ripgrep (PCRE2 compliant) is used for unopened files (opened files currently always use JS engine).
> 
> As for replace - correct replace placeholder syntax would be `$<name>` but it might be just escaped before passing to replace and should be investigated separately.
> 
> `\k<name>` is used only to reference match inside of regexp itself. `(?<start>\d+).+\k<start>` will find `123 asd 123` in `123 asd 123 asd` as expected

@IllusionMH thanks, Andrii for the clear information that is very difficult to get elsewhere. I guess the ball is with javascript engine to catch up with other languages. For now I will just use some workaround, tricks and or split into multiple regexes to prevent vscode crashes.

**Comment by pbaksa** [comment]
The problem is I want to edit the regex of a previous search in the search field, and can't guarantee that it will be valid during editing, but VSCode tries to evaluate it when it has some time because I stopped typing for thinking.
Edited regex should be re-evaluated only after enter (not the insert line break type of enter, possibly a button will be needed)

**Comment by Probmkr** [comment]
This bug is still happening in VSCode 1.67.1.

**Comment by Gloix** [comment]
Regex search (and normal search too) can take a considerable amount of time for large files, so it should never block the UI.

Can the regex search be done in a secondary thread so that it never freezes the UI? This might not help in the case that the regex search never finishes, but at least the user can keep editing the regex or copy it and then paste it after restarting the IDE.
The other option is to save the regex each time it gets executed so that it can be recovered after restarting the editor if anything fails, but I don't know if this option is too heavy on the secondary memory.

I'm usually happy with the implementation of regex search in the editor, but it has frozen the IDE more than once, and I can't say that I can safely rely on this feature.

**Comment by sancarn** [comment]
After a discussion with a friend, regexes can also freeze the syntax highlighter! 

![image](https://github.com/microsoft/vscode/assets/7938900/b9fb5237-5c07-406d-923d-538aaa1b77bf)

https://github.com/microsoft/vscode/assets/7938900/87dc3004-9b9d-474d-8858-c9b275a42108

Timeout on syntax highlighter is about 1s. The only real solution to this is to use a `DFA` based implementation of regex. [`Rust-lang`](https://github.com/rust-lang/regex/tree/master) does this. It's possible a rustic implementation could be compiled to wasm and used for regex searches and syntax highlighting. Given DFA will run in linear time, most of these performance concerns will be gone. Only issue is it won't support lookahead and behind, but for a faster experience it might be worth it.



An alternative approach

**Comment by krisutofu** [comment]
I also have this issue with the recent VSCode version.
```
Version: 1.80.1
Commit: 74f6148eb9ea00507ec113ec51c489d6ffb4b771
Date: 2023-07-12T17:22:25.257Z
Electron: 22.3.14
ElectronBuildId: 21893604
Chromium: 108.0.5359.215
Node.js: 16.17.1
V8: 10.8.168.25-electron.0
OS: Linux x64 5.4.0-153-generic snap
```

I wanted to search for a JSON key in the last object of a JSON array, this way  "...name...`(?=([^{]|\n)*\}\])`". I first had `(?=[^\{]*\}\])` which was not working for no good reason (because it is working in any modern reasonable regex engine). So I changed it by first adding parentheses `([^{])` and changing it to `([^{]|\n)` but when I only typed `|`, it immediately activated the regex search before I could type "\n" which made VSCode freeze using this suffix in the search string: `(?=([^{]|)*\}\])`. Obviously, an exponential backtracking problem of a stupid regex engine implementation, caused by the empty alternative in the inner pair of parentheses.

It is time for VSCode to get a reasonable regex engine instead of using JavaScript regex which feels so limiting in many ways. I don't even find a useful extension for better regex. Isn't there any regex engine available with cycle detection?

An intermediate solution could be to make a button to click when the regex pattern is finished so that it doesn't execute unfinished regexes.

**Comment by Json5accessible576** [comment]
@Jason3S https://openjs.space/questions/exploring-the-potential-of-vscodes-regex-search-and-replace

**Comment by EthanHonzikSPS** [comment]
Same issue here, a timeout of 5 to 10 seconds on searching would be very useful
